bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/common/QueueingThreadPoolExecutor.java

```diff
@@ -179,7 +192,15 @@ public void run() {
                                 logger.trace("Executing queued task of thread pool '{}'.", threadPoolName);
                                 QueueingThreadPoolExecutor.super.execute(runnable);
                             } else {
-                                break;
+                                try {
+                                    lock.writeLock().lock();
```

Comment:
I'm not 100% sure but shouldn't the readLock also be locked here?
Szenario: writeLock aquired here, then the Thread is paused. Switching to another Thread trying to add to the queue (line 166ff): readLock can be aquired, task added to queue, condition in line 119 is still false. When this thread continues the queuedTread is set to null and the queue again is stuck.

wdyt?

Reply:
No. See https://docs.oracle.com/javase/7/docs/api/java/util/concurrent/locks/ReentrantReadWriteLock.html

TL;DR: The locks are interconnected - while many threads can have a read-lock at the same time, the write-lock is exclusive, i.e. only one thread can have it at a time and no read-locks are possible at the same time. That's the whole clue of the ReadWriteLocks.

Reply:
Nice, thanks for the short explanation.